routing/router.go

```diff
@@ -2293,3 +2293,265 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+			if *amtToSend > capSat {
+				return fmt.Errorf("channel capacity %v "+
+					" exceeded with amt %v",
+					addr.capacity, *amtToSend)
+			}
+
+			// Check HTLC value limits.
+			if *amtToSend < addr.policy.MinHTLC ||
+				*amtToSend > addr.policy.MaxHTLC {
+
+				return errors.New("channel htlc size " +
+					"constraints violated")
+			}
+		}
+
+		// Adapt the minimum and maximum amount to send to what this
+		// channel allows.
+		//
+		// TODO(joostager): can MaxHTLC be zero?
+		if addr.policy.MinHTLC > minAmtToSend {
+			minAmtToSend = addr.policy.MinHTLC
+		}
+		if addr.policy.MaxHTLC < maxAmtToSend {
+			maxAmtToSend = addr.policy.MaxHTLC
+		}
+
+		// If we get in the situation that the minimum amount exceeds
+		// the maximum amount (enforced further down stream), we have
+		// incompatible channel policies.
+		//
+		// There is possibility with pubkey addressing that we should
+		// have selected a different channel downstream, but we don't
+		// backtrack to try to fix that.
+		if minAmtToSend > maxAmtToSend {
+			return errIncompatiblePolicies
+		}
+
+		return nil
+	}
+
+	// Traverse hops backwards to accumulate fees in the running amounts.
+	for i := len(addresses) - 1; i >= 0; i-- {
+		chanAddr := chanAddrs[i]
+		if chanAddr == nil {
+			destKey, err := btcec.ParsePubKey(
+				pubkeys[i+1][:], btcec.S256(),
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			// Iterate over candidate channels to select the channel
+			// to use for the final route.
+			fromNode := pubkeys[i]
+			toNode, err := r.cfg.Graph.FetchLightningNode(destKey)
+			if err != nil {
+				return nil, err
+			}
+
+			cb := func(tx *bbolt.Tx,
+				edgeInfo *channeldb.ChannelEdgeInfo,
+				_, inEdge *channeldb.ChannelEdgePolicy) error {
+
+				// Skip rest if found.
+				if chanAddr != nil {
+					return nil
+				}
+
+				// No unknown policy channels
+				if inEdge == nil {
+					return nil
+				}
+
+				// Before we can process the edge, we'll need to
+				// fetch the node on the _other_ end of this
+				// channel as we may later need to iterate over
+				// the incoming edges of this node if we explore
+				// it further.
+				chanFromNode, err := edgeInfo.FetchOtherNode(
+					tx, toNode.PubKeyBytes[:],
+				)
+				if err != nil {
+					return err
+				}
+
+				// Continue searching if this channel doesn't
+				// connect with the previous hop.
+				if chanFromNode.PubKeyBytes != fromNode {
+					return nil
+				}
+
+				// Validate whether this channel's policy is
+				// satisfied.
+				ca := &channelAddress{
+					policy:   inEdge,
+					capacity: edgeInfo.Capacity,
+				}
+
+				err = checkOrUpdateAmt(ca)
+				if err == nil {
+					chanAddr = ca
+				} else {
+					log.Debugf("Skipping chan %v: %v",
+						inEdge.ChannelID, err)
+				}
+
+				return nil
+			}
+
+			if err := toNode.ForEachChannel(nil, cb); err != nil {
+				return nil, err
+			}
+
+			if chanAddr == nil {
+				return nil, errors.New("no channels available")
```

Comment:
Nit: error could use some more context by including the public key.

Reply:
extended error added